astropy/table/column.py

```diff
@@ -1116,7 +1116,7 @@ def __new__(cls, data=None, name=None, mask=None, fill_value=None,
 
         if mask is None:
             if hasattr(data, 'mask'):
-                mask = data.mask
+                mask = np.array(data.mask, copy=copy)
```

Comment:
This doesn't quite go right if `data.mask is np.ma.nomask` - it makes an unnecessary mask array. Perhaps easiest to solve by rewriting the whole stanza a bit:
```
if mask is None:
    mask = getattr(data, 'mask', np.ma.nomask)

if mask is not np.ma.nomask and copy:
    mask = np.array(mask)  # copy is default; adding.
```
```
(Note: np.ma.nomask is better than `False` - it is what `MaskedArray` uses when no mask is passed in.)

Reply:
So, just to sure, this would also replace the `deepcopy` case, and the `mask = False` added by @taldcroft  recently (#7422) ? Your version makes sense but this is not the same behavior, as it results in using `nomask` where currently a full mask is created. 
Maybe there is a good reason why a full mask is created ? `nomask` is often problematic as you need to mask values with the "right way" to be sure that the mask is created when needed.

Reply:
82 failed tests (for `astropy.table`) with `nomask`, because something like `column.mask[column.data == col.null] = True` no more works.

Reply:
Ah, well, that proofs that `nomask` is indeed a bad idea (though it must be possible to trigger this in current master by passing in a masked array with `nomask`...). So, then,
```
if mask is None:
    mask = getattr(data, 'mask', False)

if mask is np.ma.nomask:
    mask = False
elif mask is not False and copy:
    mask = np.array(mask)
```

Reply:
I was trying to find a similar solution but when mask is not None, it is always copied (the `deepcopy` line), and there is a test for this.  So I end up with something like this:
```
        if mask is None:
            # Issue #7399 with fix #7422.  Passing mask=None to ma.MaskedArray
            # is extremely slow (~3 seconds for 1e7 elements), while mask=False
            # gets quickly broadcast to the expected bool array of False.
            mask = getattr(data, 'mask', False)

            if mask is not False:
                mask = np.array(data.mask, copy=copy)
        elif mask is np.ma.nomask:
            # Force the creation of a full mask array as nomask is tricky to
            # use and will fail in an unexpected manner when setting a value
            # to the mask.
            mask = False
        else:
            mask = deepcopy(mask)
```

Reply:
Looks OK to me - that even gets rid of `data.mask = nomask`.  (But it should be `np.array(mask)`, not `np.array(data.mask`).

Reply:
Good catch, thanks.